boto3/boto3/resources/base.py

```diff
@@ -0,0 +1,148 @@
+        self.data = data
+
+        # The resource model for that resource
+        self.resource_model = resource_model
+
+    def __repr__(self):
+        return 'ResourceMeta(\'{0}\', identifiers={1})'.format(
+            self.service_name, self.identifiers)
+
+    def __eq__(self, other):
+        # Two metas are equal if their components are all equal
+        if other.__class__.__name__ != self.__class__.__name__:
+            return False
+
+        return self.__dict__ == other.__dict__
+
+    def copy(self):
+        """
+        Create a copy of this metadata object.
+        """
+        params = self.__dict__.copy()
+        service_name = params.pop('service_name')
+        return ResourceMeta(service_name, **params)
+
+
+class ServiceResource(object):
+    """
+    A base class for resources.
+
+    :type client: botocore.client
+    :param client: A low-level Botocore client instance
+    """
+
+    meta = None
+    """
+    Stores metadata about this resource instance, such as the
+    ``service_name``, the low-level ``client`` and any cached ``data``
+    from when the instance was hydrated. For example::
+
+        # Get a low-level client from a resource instance
+        client = resource.meta.client
+        response = client.operation(Param='foo')
+
+        # Print the resource instance's service short name
+        print(resource.meta.service_name)
+
+    See :py:class:`ResourceMeta` for more information.
+    """
+
+    def __init__(self, *args, **kwargs):
+        # Always work on a copy of meta, otherwise we would affect other
+        # instances of the same subclass.
+        self.meta = self.meta.copy()
+
+        # Create a default client if none was passed
+        if kwargs.get('client') is not None:
+            self.meta.client = kwargs.get('client')
+        else:
+            self.meta.client = boto3.client(self.meta.service_name)
+
+        # Allow setting identifiers as positional arguments in the order
+        # in which they were defined in the ResourceJSON.
+        for i, value in enumerate(args):
+            setattr(self, '_' + self.meta.identifiers[i], value)
+
+        # Allow setting identifiers via keyword arguments. Here we need
+        # extra logic to ignore other keyword arguments like ``client``.
+        for name, value in kwargs.items():
+            if name == 'client':
+                continue
+
+            if name not in self.meta.identifiers:
+                raise ValueError('Unknown keyword argument: {0}'.format(name))
+
+            setattr(self, '_' + name, value)
+
+        # Validate that all identifiers have been set.
+        for identifier in self.meta.identifiers:
+            if getattr(self, identifier) is None:
+                raise ValueError(
+                    'Required parameter {0} not set'.format(identifier))
+
+    def __repr__(self):
+        identifiers = []
+        for identifier in self.meta.identifiers:
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

To create a `list`, try to use `list` comprehension instead of a loop. List comprehension is the preferred way to make a list using Python, and it&#39;s simpler and easier to understand than using a loop.

[Learn more](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions)